Generic Cloud Services support

MMaiero opened the pull request:
Added new apis that should simplify the integration of different cloud platforms.

As of now, this is a placeholder PR that will still see some more improvements and updates, especially for the documentation.

MMaiero committed:
New CloudConnection APIs drop, updated implementation and examples.
Added Initial contribution for the Eclipse IoT cloud connection service.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,77 @@
+     * @throws KuraException
+     *             if the connection operation fails
+     */
+    public void connect() throws KuraException;
```
Just to be concise, you could remove the access modifier from all the methods in this interface.

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,77 @@
+    public void disconnect() throws KuraException;
+
+    /**
+     * Returns true if the framework is currently connected to the remote server. The default implementation return
```
"The default implementation returns ..."

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/Constants.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * @since 2.0
+ */
+public enum Constants {
```
Doesn't it sound very generic? Couldn't we do something like `CloudConstants` or something similar?

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/Constants.java:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * The key of the property that specifies the {@code kura.service.pid} of the associated
+     * {@link CloudConnectionManager}
+     * in {@code CloudPublisher} or {@code CloudSubscriber} component configuration.
```
use {@link} instead since they are part of API and could be referenced

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/factory/CloudConnectionManagerFactory.java:
```diff
@@ -0,0 +1,112 @@
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * A {@link CloudConnectionManagerFactory} represents an OSGi Declarative Service Component
```
"OSGi DS Component" is an internal detail. I think this should not be a part of API. You could simply write - "A {@link CloudConnectionManagerFactory} is responsible for registering {@link CloudConnectionManager} ..."

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/factory/CloudConnectionManagerFactory.java:
```diff
@@ -0,0 +1,112 @@
+     * The property is set in the {@link CloudConnectionManager} instance to relate it with the Factory that generated
+     * the whole cloud stack.
+     */
+    public static final String KURA_CLOUD_CONNECTION_SERVICE_FACTORY_PID = "kura.cloud.connection.service.factory.pid";
```
Just to be concise, you could remove the access modifier from all the methods and constants from this interface.

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/factory/CloudConnectionManagerFactory.java:
```diff
@@ -0,0 +1,112 @@
+     * The IDs returned by this method must not necessarily point to registered OSGi services. But if they do, they must
+     * point only to instances of {@link CloudConnectionManager}.
+     *
+     * @return the set of services, never returns {@code null}
```
instead of never returns {@code null}, it would be more conforming to API if we write -> 

@return the set of services or an empty set

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudConnectionListener.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * {@link CloudConnectionListener} interface is implemented by applications that want to be notified on connection
+ * status changes.
+ * To be notified, the implementor need to register itself to a specific {@link CloudConnectionListenerRegistry}.
```
needs to

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudConnectionListener.java:
```diff
@@ -0,0 +1,50 @@
+import org.osgi.annotation.versioning.ConsumerType;
+
+/**
+ * {@link CloudConnectionListener} interface is implemented by applications that want to be notified on connection
```
that wants

amitjoy replied:
typo

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/request/RequestHandler.java:
```diff
@@ -0,0 +1,122 @@
+ * Implemented by services that want to implement remote resource management.
+ *
+ * <ul>
+ * <li>{@link RequestHandler#doGet} is used to implement a READ request for a resource identified by the supplied
```
the closing tag </li> is missing from all the list items

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/request/RequestHandlerConstants.java:
```diff
@@ -0,0 +1,22 @@
+package org.eclipse.kura.cloudconnection.request;
+
+/**
+ * @since 2.0
```
Since it is part of API, it would be really good for consumers to have few documentation on this class

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/request/RequestHandlerRegistry.java:
```diff
@@ -0,0 +1,28 @@
+@ProviderType
+public interface RequestHandlerRegistry {
+
+    public void register(String id, RequestHandler cloudlet) throws KuraException;
```
I believe at least few lines of javadoc would enrich this interface.

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/subscriber/CloudSubscriber.java:
```diff
@@ -0,0 +1,37 @@
+@ProviderType
+public interface CloudSubscriber extends CloudConnectionListenerRegistry {
+
+    public void register(CloudSubscriberListener listener);
```
I believe at least few lines of javadoc would enrich this interface.

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/subscriber/CloudSubscriber.java:
```diff
@@ -0,0 +1,37 @@
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * Interface intended to have a specific implementation associated to a {@link CloudConnectionManager} and with
```
is intended

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/subscriber/listener/CloudSubscriberListener.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * {@link CloudSubscriberListener} is the interface to be implemented by applications that needs to be notified of
+ * events in the subscriber.
+ * Arrived methods are invoked whenever a message is received in the associated {@link CloudSubscriber}.
```
I think "Arrived" sounds a bit odd.

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/subscriber/listener/CloudSubscriberListener.java:
```diff
@@ -0,0 +1,39 @@
+ * @since 2.0
+ */
+@ConsumerType
+public interface CloudSubscriberListener {
```
Add @FunctionalInterface

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/subscriber/package-info.java:
```diff
@@ -0,0 +1,14 @@
+ *******************************************************************************/
+/**
+ * Provides services for subscribing to remote servers and allows to invoke local applications.
+ *
```
Extra line

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/subscriber/listener/package-info.java:
```diff
@@ -0,0 +1,14 @@
+ *******************************************************************************/
+/**
+ * Provides apis that allow to register subscriber applications.
+ *
```
Extra line

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/KuraAlertPayload.java:
```diff
@@ -0,0 +1,120 @@
+
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
```
The intention of the class could be specified in javadoc

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/KuraAlertSeverity.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @since 2.0
+ */
+public enum KuraAlertSeverity {
```
The javdoc would be helpful for consumers

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/KuraAlertStatus.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @since 2.0
+ */
+public enum KuraAlertStatus {
```
The javadoc could make the enum more comprehensive

amitjoy commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudClientListenerAdapter.java:
```diff
@@ -0,0 +1,91 @@
+
+    private static final String IGNORED_ERROR_NOTIFYING = "IGNORED - Error notifying ";
+
+    private static final Logger s_logger = LoggerFactory.getLogger(CloudClientListenerAdapter.class);
```
logger

amitjoy commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudClientListenerAdapter.java:
```diff
@@ -0,0 +1,91 @@
+
+    private static final Logger s_logger = LoggerFactory.getLogger(CloudClientListenerAdapter.class);
+
+    private final CloudClientListener m_listener;
```
listener

amitjoy commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudConnectionManagerImpl.java:
```diff
@@ -0,0 +1,581 @@
+    private PositionService positionService;
+    private EventAdmin eventAdmin;
+
+    // package visibility for LyfeCyclePayloadBuilder
```
Life

amitjoy commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadEncoder.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Common interface for the PayloadEncoders
+ */
+public interface CloudPayloadEncoder {
```
@FunctionalInterface

amitjoy commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadProtoBufDecoderImpl.java:
```diff
@@ -0,0 +1,162 @@
+
+public class CloudPayloadProtoBufDecoderImpl {
+
+    private static final Logger s_logger = LoggerFactory.getLogger(CloudPayloadProtoBufDecoderImpl.class);
```
logger

amitjoy commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadProtoBufDecoderImpl.java:
```diff
@@ -0,0 +1,162 @@
+
+    private static final Logger s_logger = LoggerFactory.getLogger(CloudPayloadProtoBufDecoderImpl.class);
+
+    private byte[] m_bytes;
```
bytes

amitjoy commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/message/KuraDeviceProfile.java:
```diff
@@ -0,0 +1,870 @@
+ * The KuraDeviceProfile is a container class that holds the parameters that make up the from of a device.
+ * This information is used to build the birth and disconnect certificates that are published when
+ * connecting to and disconnecting from the broker.
+ *
```
extra line

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/publisher/CloudPublisherImpl.java:
```diff
@@ -0,0 +1,289 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CloudPublisherImpl
```
It seems that the Eclipse IoT CloudPublisher does not publish the "apps certificate".
As a consequence, Kapua will not list the application identifiers.

MMaiero replied:
The apps certificate works with this refactoring only for "Cloudlets". With the new concept of Publisher and the concept of multiple applications bound to a single publisher it will have no sense.
We definitely need to update the basic publisher modifying the concept of appId that could be misleading.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/publisher/CloudPublisherImpl.java:
```diff
@@ -0,0 +1,289 @@
+
+    @Override
+    public void onDisconnected() {
+        this.cloudConnectionListeners.forEach(CloudConnectionListener::onDisconnected);
```
iterate catching exceptions thrown by listeners

MMaiero replied:
I believe the same is also for existing cloud clients, right? Or I'm missing something...

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/publisher/CloudPublisherImpl.java:
```diff
@@ -0,0 +1,289 @@
+
+    @Override
+    public void onConnectionLost() {
+        this.cloudConnectionListeners.forEach(CloudConnectionListener::onConnectionLost);
```
same

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/publisher/CloudPublisherImpl.java:
```diff
@@ -0,0 +1,289 @@
+
+    @Override
+    public void onConnectionEstablished() {
+        this.cloudConnectionListeners.forEach(CloudConnectionListener::onConnectionEstablished);
```
same

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/publisher/CloudPublisherImpl.java:
```diff
@@ -0,0 +1,289 @@
+            index++;
+        }
+
+        this.cloudDeliveryListeners.forEach(deliveryListener -> deliveryListener.onMessageConfirmed(messageId));
```
same as above

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/metatype/org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher.xml:
```diff
@@ -0,0 +1,46 @@
+         description="The CloudPublisher allows to define publishing parameters and provide a simple endpoint where the applications can attach to publish their messages.">
+        
+        <AD id="semantic.topic"
+            name="Semantic Topic"
```
I've checked and "Semantic Topic" is OK as this term is also used in Kapua (KapuaChannel.getSemanticParts()). In Hono, the semantic topic becomes the MQTT topic suffix. You can ignore this comment

MMaiero replied:
So?

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/metatype/org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher.xml:
```diff
@@ -0,0 +1,46 @@
+            cardinality="0"
+            required="false"
+            default="W1/A1/$assetName"
+            description='Specifies the semantic topic.'>
```
”The MQTT topic suffix providing the interpretation of the message.”

MMaiero replied:
Agreed

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/metatype/org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher.xml:
```diff
@@ -0,0 +1,46 @@
+            description="Type of message to be published.">
+            <Option label="Telemetry QoS 0" value="telemetryQos0" />
+            <Option label="Telemetry QoS 1" value="telemetryQos1" />
+            <Option label="Events" value="events" />
```
"Event", without 's'

MMaiero replied:
In the official reference used it is named as Events: https://docs.google.com/document/d/1ZUv5ux8QLcv9vTwtFuJLC6gB0vo6-OzLwSZXcVFQLUI/edit#heading=h.z213w17kspox

But I'll change it

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/metatype/org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher.xml:
```diff
@@ -0,0 +1,46 @@
+            type="String"
+            cardinality="0"
+            required="true"
+            default="telemetryQos0"
```
An integer would be enough and would waste less space and bandwidth.
However, using integers is probably error prone. In this case I would at least use String values matching Sonar Lint patterns for enums, for example TELEMETRY_QOS0 (and define the enum)

MMaiero replied:
https://github.com/eclipse/kura/blob/9b4efdb8ec6c77277cc7f951c8f94647b5deafff/kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/publisher/MessageType.java

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/metatype/org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher.xml:
```diff
@@ -0,0 +1,46 @@
+            type="String"
+            cardinality="0"
+            required="false"
+            default="W1/A1/$assetName"
```
1. Still not convinced about this default which is only suitable for a Wire Publisher.

2. Please explain succintly explain how the variable substitution works in the field description

MMaiero replied:
As described in one of the previous comments, this topic supports wires publishing out of the box.
It's obvious that it does not support all the cases, but it is an example that can be used to describe the structure and the substitutions.

When the user creates the publisher, the web ui presents the full configuration and, at that point, the user can decide to specify the topic, depending on his needs.
I believe that a small minority of the overall publishers will stay with the default configuration.

An option could be to specify an empty or not meaningful topic and have the wires default defined in the comment. But, personally, this is an option I don't like much.

I completely agree that we should improve the description, taking care to specify how substitutions work.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/metatype/org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher.xml:
```diff
@@ -0,0 +1,46 @@
+    <OCD id="org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher" 
+         name="CloudPublisher" 
+         description="The CloudPublisher allows to define publishing parameters and provide a simple endpoint where the applications can attach to publish their messages.">
+        
```
"The Eclipse IoT Cloud Publisher provides a service to publish event and telemetry messages to a cloud platform compatible with the Eclipse IoT MQTT topic namespace."

MMaiero replied:
Slightly modified, but included.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/metatype/org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher.xml:
```diff
@@ -0,0 +1,46 @@
+            <Option label="Telemetry QoS 0" value="telemetryQos0" />
+            <Option label="Telemetry QoS 1" value="telemetryQos1" />
+            <Option label="Events" value="events" />
+            <Option label="Alerts" value="alerts" />
```
Not yet implemented in the Eclipse IoT integration (by Hono). Should we remove it and add it later on?

MMaiero replied:
I don't know. In any case the bundle will see more updates as far as the EclipseCON deadline approaches. An eventual cleanup will be done later on in any case.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,36 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: org.eclipse.kura.cloud.mqtt.eclipseiot
+Bundle-SymbolicName: org.eclipse.kura.cloud.mqtt.eclipseiot;singleton:=true
```
As the API package is `org.eclipse.kura.cloudconnection` we should use `org.eclipse.kura.cloudconnection.eclipseiot.provider` for the project name and the BSN. This is the convention we have followed for `Driver`s and `Asset`s.

`Bundle-Name` could be more readable, for example `Connection Provider for the Eclipse IoT Cloud Platform`

MMaiero replied:
Agreed.
Not sure about removing mqtt from the name. I'm expecting that every implementation will be in a separate bundle. So, this one will take care of mqtt. In future, maybe, there will be another protocol and that could be coded in the name as well

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/cloud.xml:
```diff
@@ -0,0 +1,62 @@
+              interface="org.osgi.service.event.EventAdmin"/>
+   <property name="kura.ui.service.hide" type="Boolean" value="true"/>
+   <property name="kura.ui.factory.hide" type="String" value="true"/>
+   <property name="service.pid" type="String" value="org.eclipse.kura.cloud.mqtt.eclipseiot.CloudConnectionManager"/>
```
1. Do we really need to explicitly set the `service.pid` property? I believe `component.name` should be enough.
2. `org.eclipse.kura.cloud.mqtt.eclipseiot.CloudConnectionManager` is what is shown in the Web UI as the Factory PID of the Cloud Connection. However, I believe that a shorter name like `org.eclipse.kura.cloud.mqtt.eclipseiot` would work better.
3. Based on the comment above to the name of the project and BSN, it would be nice if the Factory PID of a Cloud Connection followed the scheme `org.eclipse.kura.cloudconnection.${provider-name}`. In this case `org.eclipse.kura.cloudconnection.eclipseiot`.
The same convention should be also applied to the "classic" Cloud Connection, for example `org.eclipse.kura.cloudconnection.classic`.

MMaiero replied:
1. No it is not needed here, since completed by the factory. But was there since it was copied from the original cloudService implementation. I'll update both.

MMaiero replied:
2. org.eclipse.kura.cloud.mqtt.eclipseiot.CloudConnectionManager(-something) is the actual kura.service.pid required by the factory. CloudConnectionManager(-something) will be used by the web UI to display the proper stack of components for the cloud service. I'll try to see if eclipseiot(-something) will work as well.

But personally having a tab called classic(-something) does not make me so enthusiastic.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/cloudPublisher.xml:
```diff
@@ -0,0 +1,25 @@
+      <provide interface="org.eclipse.kura.configuration.ConfigurableComponent"/>
+   </service>
+   <property name="service.pid" type="String" value="org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher"/>
+   <property name="cloud.connection.service.factory.pid" type="String" value="org.eclipse.kura.cloud.mqtt.eclipseiot.CloudConnectionManager"/>
```
`cloud.connection.service.factory.pid`: the word `service` is redundant. Is this more consistent: `cloudconnection.factory.pid`?

MMaiero replied:
It's an internal property used only by the web ui. We can rename it without issues.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/OSGI-INF/cloudPublisher.xml:
```diff
@@ -0,0 +1,25 @@
+      <provide interface="org.eclipse.kura.cloudconnection.publisher.CloudPublisher"/>
+      <provide interface="org.eclipse.kura.configuration.ConfigurableComponent"/>
+   </service>
+   <property name="service.pid" type="String" value="org.eclipse.kura.cloud.mqtt.eclipseiot.CloudPublisher"/>
```
As described below, we should establish a convention for the PID of a publisher and subscriber providers.
For example `org.eclipse.kura.cloudconnection.{provider-name}.{protocol}.pub|sub`. In this case they would be `org.eclipse.kura.cloudconnection.eclipseiot.mqtt.pub` and `org.eclipse.kura.cloudconnection.eclipseiot.mqtt.sub`

MMaiero replied:
This could be done, but only at web ui level (needs specific code).
It could not be forced in any way for custom providers.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/KuraAlertStatus.java:
```diff
@@ -0,0 +1,19 @@
+/*******************************************************************************
```
We should spend more thoughts on the alert status.

We have mixed the concept of "alert", an event occurrence, with that of "alarm" which represents a state and which can be reset.

Example:
Temperature goes above threshold: an alert is sent with an x code.
The cloud platform receives the alert and logs it in a database.

The platform creates an alarm if it does not already exists.
The alarm can be reset by the user.

The temperature goes below threshold: an alert is sent with a code y.
The platform clears the alarm, if it exists

In short, the alarms exist only in the platform and the alerts have no status.

Therefore:
It is our task to define x and y codes for alerts.
It is the task of a businnes logic to implement alarms.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudClientListenerAdapter.java:
```diff
@@ -0,0 +1,91 @@
+ * Contributors:
+ *     Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.cloud.mqtt.eclipseiot.internal.cloud;
```
Based on our practices, I believe `internal` should immediately follow `org.eclipse.kura`.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudClientListenerAdapter.java:
```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2016 Eurotech and/or its affiliates
```
Date

MMaiero replied:
All those files are simple copies made from the original bundle. No changes made.
In terms of copyright it is still not clear to me if they need to be considered new file (only 2018) or need to keep the original header (maybe the end year needs to be updates)

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudClientListenerAdapter.java:
```diff
@@ -0,0 +1,91 @@
+        try {
+            this.listener.onControlMessageArrived(deviceId, appTopic, msg, qos, retain);
+        } catch (Exception e) {
+            logger.error(IGNORED_ERROR_NOTIFYING + this.listener + " for onControlMessageArrived", e);
```
Is string concatenation more efficient?

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadGZipEncoder.java:
```diff
@@ -0,0 +1,34 @@
+
+public class CloudPayloadGZipEncoder implements CloudPayloadEncoder {
+
+    private final CloudPayloadEncoder m_decorated;
```
Remove "m_".

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadGZipEncoder.java:
```diff
@@ -0,0 +1,34 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2016 Eurotech and/or its affiliates
```
Date. This file seems a copy. Cannot we move it to an utils bundle?

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,36 @@
+ org.eclipse.kura.cloudconnection.publisher;version="[1.0,1.1)",
+ org.eclipse.kura.configuration;version="[1.1,2.0)",
+ org.eclipse.kura.configuration.metatype;version="[1.1,2.0)",
+ org.eclipse.kura.core.data;version="[1.0,2.0)",
```
Internal package

MMaiero replied:
True, but needed to get the connection information from the DataService. The alternative is to change the API

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,36 @@
+ org.eclipse.kura.cloudconnection.publisher;version="[1.0,1.1)",
+ org.eclipse.kura.configuration;version="[1.1,2.0)",
+ org.eclipse.kura.configuration.metatype;version="[1.1,2.0)",
+ org.eclipse.kura.core.data;version="[1.0,2.0)",
```
Internal package

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,36 @@
+ org.eclipse.kura.configuration;version="[1.1,2.0)",
+ org.eclipse.kura.configuration.metatype;version="[1.1,2.0)",
+ org.eclipse.kura.core.data;version="[1.0,2.0)",
+ org.eclipse.kura.core.util;version="[1.0,2.0)",
```
Internal package

MMaiero replied:
True, but needed for a set of utilities used also by the "classic" stack we provide.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,36 @@
+ org.eclipse.kura.configuration;version="[1.1,2.0)",
+ org.eclipse.kura.configuration.metatype;version="[1.1,2.0)",
+ org.eclipse.kura.core.data;version="[1.0,2.0)",
+ org.eclipse.kura.core.util;version="[1.0,2.0)",
```
Internal package

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/build.properties:
```diff
@@ -0,0 +1,13 @@
+               OSGI-INF/,\
+               about.html,\
+               about_files/,\
+               lib/com.google.protobuf_2.6.0.jar
```
Oh, no!

MMaiero replied:
I know, but at the moment the other alternative I see is a rewrap.

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/build.properties:
```diff
@@ -0,0 +1,13 @@
+               OSGI-INF/,\
+               about.html,\
+               about_files/,\
+               lib/com.google.protobuf_2.6.0.jar
```
Oh, no!

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadJsonEncoder.java:
```diff
@@ -0,0 +1,145 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Eurotech and/or its affiliates
```
As above

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadJsonDecoder.java:
```diff
@@ -0,0 +1,162 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Eurotech and/or its affiliates
```
Date. Another copy?

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadJsonFields.java:
```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Eurotech and/or its affiliates
```
Same as above

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadProtoBufDecoderImpl.java:
```diff
@@ -0,0 +1,162 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2016 Eurotech and/or its affiliates
```
Same as above

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudPayloadProtoBufEncoderImpl.java:
```diff
@@ -0,0 +1,161 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2017 Eurotech and/or its affiliates and others
```
Same as above

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudServiceLifecycleCertsPolicy.java:
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Eurotech and/or its affiliates
```
Date

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudServiceOptions.java:
```diff
@@ -0,0 +1,291 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2017 Eurotech and/or its affiliates
```
Date, copied and there's no Cloud Service here

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudServiceOptions.java:
```diff
@@ -0,0 +1,291 @@
+import java.util.Map;
+
+import org.eclipse.kura.cloud.CloudPayloadEncoding;
+import org.eclipse.kura.core.util.ProcessUtil;
```
Why? Can you use Apache Commons Exec?

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudServiceOptions.java:
```diff
@@ -0,0 +1,291 @@
+            // Try to get the device hostname
+            else if (deviceDisplayNameOption.equals("hostname")) {
+                displayName = "UNKNOWN";
+                if (SystemService.OS_MAC_OSX.equals(this.systemService.getOsName())) {
```
All the platform specific code should be moved to the System Service that should return the device name no matter how obtained. At least all the crap will be there. Do this also for the "classic" Cloud Connection

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/CloudServiceOptions.java:
```diff
@@ -0,0 +1,291 @@
+        return LIFECYCLE_RETAIN;
+    }
+
+    private String getHostname(String command) {
```
See above

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/LifeCyclePayloadBuilder.java:
```diff
@@ -0,0 +1,212 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2018 Eurotech and/or its affiliates
```
Copied. It seems that the classic and Eclipse IoT providers should use a common service for all the things they share

cdealti commented on kura/org.eclipse.kura.cloud.mqtt.eclipseiot/src/main/java/org/eclipse/kura/cloud/mqtt/eclipseiot/internal/cloud/LifeCyclePayloadBuilder.java:
```diff
@@ -0,0 +1,212 @@
+
+    private static final String UNKNOWN = "UNKNOWN";
+
+    private final CloudConnectionManagerImpl cloudServiceImpl;
```
`cloudServiceImpl` there's no such thing here

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/publisher/CloudNotificationPublisher.java:
```diff
@@ -0,0 +1,25 @@
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * Publisher API to be used in {@link RequestHandler} implementations that need to send notification messages to remote
```
<pre>Marker interface used to identify a Cloud Publisher dedicated to send notifications until completion of a long-lived operation initially handled by a {@link RequestHandler}.</pre>

Please, also explain how to identify a specific notification publisher

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/publisher/CloudNotificationPublisher.java:
```diff
@@ -0,0 +1,25 @@
+ * @noimplement This interface is not intended to be implemented by clients.
+ */
+@ProviderType
+public interface CloudNotificationPublisher extends CloudPublisher {
```
Maybe just `NotificationPublisher`

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/publisher/CloudPublisher.java:
```diff
@@ -0,0 +1,96 @@
+import org.eclipse.kura.cloudconnection.message.KuraMessage;
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
```
Javadoc no longer accurate

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/publisher/CloudPublisher.java:
```diff
@@ -0,0 +1,96 @@
+public interface CloudPublisher {
+
+    /**
+     * Publishes the received {@link KuraMessage} using the associated Cloud Stack.
```
We don't have a concept of a "cloud stack". Maybe just
`Publishes a {@link KuraMessage}`

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/publisher/CloudPublisher.java:
```diff
@@ -0,0 +1,96 @@
+     *
+     * @param message
+     *            The {@link KuraMessage} to be published
+     * @return a String representing the message ID
```
`null` is returned if the Cloud Publisher will not confirm the message delivery, either because this is not supported by the underlying protocol or the Cloud Publisher itself is not implemented or configured to request the confirmation.

cdealti commented on kura/examples/org.eclipse.kura.example.publisher/OSGI-INF/metatype/org.eclipse.kura.example.publisher.ExamplePublisher.xml:
```diff
@@ -19,22 +19,22 @@
-        <AD id="cloud.service.pid"
-            name="cloud.service.pid"
+        <AD id="CloudPublisher.target"
+            name="CloudPublisher.target"
```
If we make the AD names more readable, I suggest "CloudPublisher Target Filter"

cdealti commented on kura/examples/org.eclipse.kura.example.publisher/OSGI-INF/metatype/org.eclipse.kura.example.publisher.ExamplePublisher.xml:
```diff
@@ -19,22 +19,22 @@
-            default="org.eclipse.kura.cloud.CloudService"
-            description="Specify the pid of the Cloud Service that has to used to publish messages to the cloud platform.">
+            default="(kura.service.pid=changeme)"
+            description="Specify the pid of the Cloud publisher that has to used to publish messages to the cloud platform.">
```
Specifies the PID of the target Cloud Publisher as an OSGi target filter.

cdealti commented on kura/examples/org.eclipse.kura.example.publisher/OSGI-INF/metatype/org.eclipse.kura.example.publisher.ExamplePublisher.xml:
```diff
@@ -19,22 +19,22 @@
-            default="EXAMPLE_PUBLISHER"
-            description="Application ID">
+            required="false"
+            default=""
```
why empty default for the subscriber?

cdealti commented on kura/examples/org.eclipse.kura.example.publisher/OSGI-INF/metatype/org.eclipse.kura.example.publisher.ExamplePublisher.xml:
```diff
@@ -19,22 +19,22 @@
-            description="Application ID">
+            required="false"
+            default=""
+            description="Specify the pid of the Cloud subscriber that has to used to receive messages from the cloud platform.">
```
See above for the suggested changes for the publisher

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/KuraErrorCode.java:
```diff
@@ -105,5 +105,17 @@
+    /**
+     * @since 2.0
+     */
+    BAD_REQUEST,
```
Kura documentation reports the following error codes

200 (RESPONSE_CODE_OK)
400 (RESPONSE_CODE_BAD_REQUEST)
404 (RESPONSE_CODE_NOTFOUND)
500 (RESPONSE_CODE_ERROR)

1. I don't see the enum for RESPONSE_CODE_ERROR but maybe another error code is used for it
2. I would prefix the new enums with `RESPONSE_`

[1] https://eclipse.github.io/kura/ref/mqtt-namespace.html

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloud/CloudClientListener.java:
```diff
@@ -22,8 +24,11 @@
  * The Arrived method signatures are differentiated based on whether the incoming messages have been
  * published to a data topic (by default accountName/#) or a control topic (by default $EDC/accountName/#).
+ * 
+ * @deprecated Please use {@link CloudConnectionListener} and {@link CloudSubscriberListener} instead
```
Be consistent with previous text: "Please consider using XXX"

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionConstants.java:
```diff
@@ -0,0 +1,48 @@
+
+    private String value;
+
+    CloudConnectionConstants(final String value) {
```
default method visibility OK?

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionConstants.java:
```diff
@@ -0,0 +1,48 @@
+import org.eclipse.kura.cloudconnection.subscriber.CloudSubscriber;
+
+/**
+ * This enum provides constants that can be used by cloud connection implementations.
```
I believe these constants are relevant not only to implementations if, by implementations you mean Cloud Connection providers.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionConstants.java:
```diff
@@ -0,0 +1,48 @@
+     * {@link CloudConnectionManager}
+     * in {@link CloudPublisher} or {@link CloudSubscriber} component configuration.
+     */
+    CLOUD_CONNECTION_SERVICE_PID_PROP_NAME("cloud.connection.service.pid"),
```
TODO review the property names after the below comment about the CloudConnectionManager.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * The {@link CloudConnectionManager} provides an API layer to ease the communication with a remote server.
```
I don't agree. It's only an API to manage the connection to a server and assumes a protocol that is connection oriented and using long-lived/always on connections.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * The {@link CloudConnectionManager} provides an API layer to ease the communication with a remote server.
+ * {@link CloudConnectionManager} abstracts the developers from the complexity of the transport protocol and payload
+ * format used in the communication.
```
You are not talking about this API.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+ * format used in the communication.
+ * It allows for a single connection to a remote server to be shared across more than one {@link CloudPublisher} and
+ * {@link CloudSubscriber} running in the framework.
+ *
```
Again, you are not talking about this API.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+ *
+ * Each cloud provider should provide its own {@link CloudConnectionManager} implementation in order to best fit the
+ * specific requirements.
+ *
```
Redundant.
The intention of this API is clear: connect and disconnect. It's obvious that the details depend on the provider.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+ *
+ * Applications should not use directly this API but, instead, use the {@link CloudPublisher} and the
+ * {@link CloudSubscriber} interfaces to give applications the capabilities to publish and receive messages.
+ *
```
Why not? Obviously it does not make sense that the connection is managed, without coordination, by multiple consumers that are not aware of each other.
Typically, there's a single consumer that will manage the connection, for example the Kura Web UI but there's nothing preventing another one doing the same.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+ *
+ * This service interface MAY be optionally registered by those protocols that need to establish a long-lived
+ * connection.
+ *
```
I interpret this last comment as "Cloud Connection providers may **not** register this service interface".

However this would contradict the Javadoc of `CloudConnectionManagerFactory`:

> A {@link CloudConnectionManagerFactory} is responsible to register {@link CloudConnectionManager} instances in the framework.

I my head a Cloud Connection provider is any "implementation" that **may** register
1. `CloudPublisher`S (0 or more implementations)
2. `CloudSubscribers`S (0 or more implementations)
3. `RequestHandler`S (0 or more implementations)
4. `CloudConnectionManager` (at most one implementation)
5. A factory service interface to create and delete Cloud Connection instances (exactly one implementation per provider).

As a consequence, it would be entirely possible to implement a Cloud Connection that does not provide any of the items 1 to 4 above, even if such implementation would not be very useful.

If all this is possible, naming the above factory `CloudConnectionManagerFactory` is misleading as it is not required to provide a `CloudConnectionManager` service to the framework.

I believe that a better name would be `CloudConnectionFactory`. As all the items 1 to 4 above are optional, what is `CloudConnectionFactory` supposed to do when it's asked to create a new Cloud Connection? One possibility is to register a new marker interface, to allow identifying a Cloud Connection as a whole.  

It comes to no surprise that this marker interface will be named `CloudConnection`.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+
+/**
+ * The {@link CloudConnectionManager} provides an API layer to ease the communication with a remote server.
+ * {@link CloudConnectionManager} abstracts the developers from the complexity of the transport protocol and payload
```
It should be enough to link a API only once in the Javadoc

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/factory/CloudConnectionManagerFactory.java:
```diff
@@ -0,0 +1,112 @@
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * A {@link CloudConnectionManagerFactory} is responsible to register {@link CloudConnectionManager} instances in the
```
See comment above about `CloudConnectionManager`.
TODO: I will review this again based on that discussion.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+public interface CloudConnectionManager {
+
+    /**
+     * Connects the framework to the remote cloud server.
```
"Establishes a connection to the cloud platform configured in the associated Cloud Connection"

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Connects the framework to the remote cloud server.
+     *
+     * @throws KuraException
```
I believe we already have a specific KuraException for this

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+     * Connects the framework to the remote cloud server.
+     *
+     * @throws KuraException
+     *             if the connection operation fails
```
"If the connection fails" or "if the operation fails"

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+    public void connect() throws KuraException;
+
+    /**
+     * Performs a safe disconnection from the remote server.
```
"Performs a clean disconnection from the cloud platform."

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Performs a safe disconnection from the remote server.
+     *
+     * @throws KuraException
```
Same as for `connect`

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+    public void disconnect() throws KuraException;
+
+    /**
+     * The default implementation returns true if the framework is currently connected to the remote server. The default
```
"Tests if the connection is alive"

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return {@code true} if the framework is connected to the remote server. {@code false} otherwise.
+     */
+    public default boolean isConnected() {
```
Why do we need a default implementation?

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Provides information related to the associated connection. The information provided depends on the specific
+     * implementation and type of connection to the remote resource. The default implementation returns an empty
+     * {@link Map}
```
Is there any well known property already defined that a consumer should know?

P.S. You don't need to provide a link to Map

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/CloudConnectionManager.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    /**
+     * The implementation will register the {@link CloudConnectionListener} instance passed as argument. Once a cloud
```
Registers listeners of connection-related occurrences.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudConnectionListener.java:
```diff
@@ -0,0 +1,49 @@
+import org.osgi.annotation.versioning.ConsumerType;
+
+/**
+ * The {@link CloudConnectionListener} interface is implemented by applications that want to be notified on connection
```
Do not provide a link to ourselves ;)

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudConnectionListener.java:
```diff
@@ -0,0 +1,49 @@
+public interface CloudConnectionListener {
+
+    /**
+     * Notifies that the cloud stack has cleanly disconnected.
```
Can we avoid the term "cloud stack"?

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudConnectionListener.java:
```diff
@@ -0,0 +1,49 @@
+    public void onDisconnected();
+
+    /**
+     * Called when the client has lost its connection with the broker. This is only a notification, the callback handler
```
This has been probably copied from elsewhere but so far we have tried to be protocol agnostic and avoided mentioning clients and broker.

I believe it's enough to say that "The connection to the cloud platform has been dropped."

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudConnectionListener.java:
```diff
@@ -0,0 +1,49 @@
+     * Called when the client has lost its connection with the broker. This is only a notification, the callback handler
+     * should not attempt to handle the reconnect.
+     *
+     * If the bundle using the client relies on subscriptions beyond the default ones,
```
This information is (only) useful to a `CloudSubscriber` implementation but this already knows all the implementation details of its Cloud Connection.

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudConnectionListener.java:
```diff
@@ -0,0 +1,49 @@
+    public void onConnectionLost();
+
+    /**
+     * Called when the cloud stack has successfully connected to the broker.
```
Similar considerations as above

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudDeliveryListener.java:
```diff
@@ -0,0 +1,34 @@
+import org.osgi.annotation.versioning.ConsumerType;
+
+/**
+ * Implementors of this interface will be able to handle cloud stack related
```
1. Can we avoid the "cloud stack" term?
2. Can we avoid the "event" term an use "acknowledge"

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/CloudDeliveryListener.java:
```diff
@@ -0,0 +1,34 @@
+public interface CloudDeliveryListener {
+
+    /**
+     * Confirms message delivery to the remote cloud platform.
```
I believe "remote" is redundant

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/listener/package-info.java:
```diff
@@ -0,0 +1,14 @@
+ *
+ *******************************************************************************/
+/**
+ * Provides apis that allow to listen to cloud connection events.
```
Maybe: "Provides APIs to listen to cloud connection notifications, such as connection status changes and message delivery confirmations."

cdealti commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloudconnection/message/KuraMessage.java:
```diff
@@ -0,0 +1,55 @@
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * The class KuraMessage represents a message that is shared between an application running in the framework and a
```
This concept deserves better documentation.
The most obvious assumption made by the API consumer is that `KuraPayload` will be eventually serialized in the payload of the underlying protocol. For example, with the Eclipse IoT Cloud Connection provider, Kura Payload can be encoded to a Kura Protobuf or to the "Simple JSON" format and the bytes placed in the MQTT.

Even in this simple case, things are not so simple as the encoding can be either lossless (Kura protobuf) or lossy (Simple JSON).

Another concept is the semantics of the message. An application using Cloud Publisher assumes that the  semantics (if any) of the message is added by the Cloud Publisher itself, for example through the "semantic topic".

So the question is what is the role of the properties and in what cases applications should use them.

The first use case is that of a shared Cloud Publisher used by multiple alike applications.
For example, Wire Publisher is a `WireComponent` converting `WireEnvelope`S to `KuraMessage`S and publishing them through a Cloud Publisher. There might be multiple Wire Publishers in a graph sharing the same Cloud Publisher and this can be configured in a way to build the full semantics of the message based on its configuration and a message property.

Example
Kura and Kapua define a namespace for messages published on behalf of `Asset`S.
Over MQTT this namespace is represented by the MQTT semantic topic W1/A1/$assetName.
Kapua must be able to consume the messages published in this namespace without having to decode the MQTT payload. Kapua can efficiently consume the whole namespace or only the messages published by a specific Asset.
Wire Publishers, publishing on behalf of Assets, will add a property `assetName` to the properties of the Kura Message and publish it through a Cloud Publisher whose semantic topic is configured as shown above.
The associated Cloud Publisher will interpret `$assetName` as the value of the property `assetName`.

I believe this rationale must be given somewhere in the Javadoc, possibly in a package info, not sure which one.

cdealti reviewed (changes requested):
Reviewing such a big change is difficult. It introduces a new API but it also refactors the existing code to leverage the new API. I should have stopped the review earlier and I haven't even look at all the changes. However, I suggest to consider first the comments to the API, starting from those to `CloudConnectionManager`. Another important comment is to `KuraMessage`.
In general we should try to use in the Javadoc a consistent terminology to explain things.

MMaiero committed:
Updated messages in CommandCloudApp

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Fixes in wire publisher.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Updated the examples to leverage the new APIs.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Solved missing imports in examples and failing tests.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Implemented Command and Control in eclipseiot provider.

This matches the changes done in Eclipse Hono.
Note: to work well, the mqtt maximum message size should be increased in the Hono instance.

Updated the KuraTopic API. This is a breaking change.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Added TODOs to reference parts that are still missing in eclipseiot bundle.

Cleaned kura.buils.props file.
Cleaned stale imports.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Fixed tests.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Completed switch to compressed topic format.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Updated javadocs.

Minor typo fixes.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>
